### PR TITLE
[Schema change] Support More column type in schema change

### PR DESCRIPTION
**Pull request by HappenLee**
1. Support modify column type CHAR to TINYINT/SMALLINT/INT/BIGINT/LARGEINT/FLOAT/DOUBLE/DATE
and TINYINT/SMALLINT/INT/BIGINT/LARGEINT/FLOAT/DOUBLE convert to a wider range of numeric types (#4937)

2. Use template to refactor code of types.h and schema_change.cpp to delete redundant code.

fix issue:#4937

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #ISSUE), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

File: be/src/olap/types.h [hunk 1]
```diff
@@ -495,12 +496,32 @@ struct BaseFieldtypeTraits : public CppTypeTraits<field_type> {
...
+
 template <typename T>
-OLAPStatus convert_int_from_varchar(void* dest, const void* src) {
-    using SrcType = typename CppTypeTraits<OLAP_FIELD_TYPE_VARCHAR>::CppType;
-    auto src_value = reinterpret_cast<const SrcType*>(src);
+T convert_from_varchar(const Slice* src_value, StringParser::ParseResult& parse_res, std::true_type) {
+    return StringParser::string_to_int<T>(src_value->get_data(), src_value->get_size(), &parse_res);
+}
+
+template <typename T>
```
Comment (chaoyli) [inline be/src/olap/types.h:514]: There may use traits ?
Or use if constexpr (std::is_floating_point_v<T>)) to distinguish the two cases.

File: be/src/olap/types.h [hunk 2]
```diff
@@ -509,98 +530,55 @@ OLAPStatus convert_int_from_varchar(void* dest, const void* src) {
...
-    static OLAPStatus convert_from(void* dest, const void* src, const TypeInfo* src_type, MemPool* mem_pool) {
-        if (src_type->type() == OLAP_FIELD_TYPE_VARCHAR) {
-            return convert_int_from_varchar<CppType>(dest, src);
-        }
-        return OLAPStatus::OLAP_ERR_INVALID_SCHEMA;
-    }
-};
+template <FieldType fieldType>
+struct FieldTypeTraits : public ArithmeTicFieldtypeTraits<fieldType,
+        std::is_arithmetic<typename BaseFieldtypeTraits<fieldType>::CppType>::value && std::is_signed<typename BaseFieldtypeTraits<fieldType>::CppType>::value> {};
```
Comment (chaoyli) [inline be/src/olap/types.h:563]: use std::is_signed for what?

File: be/src/olap/types.h [hunk 3]
```diff
@@ -509,98 +530,55 @@ OLAPStatus convert_int_from_varchar(void* dest, const void* src) {
...
+    prepare_char_before_convert(src);
+    return arithmetic_convert_from_varchar<T>(dest, src);
 }
 
-template<FieldType field_type>
-struct FieldTypeTraits : public BaseFieldtypeTraits<field_type> { };
+// Using ArithmeTicFieldtypeTraits to Derived code for OLAP_FIELD_TYPE_XXXINT, OLAP_FIELD_TYPE_FLOAT,
+// OLAP_FIELD_TYPE_DOUBLE, to reduce redundant code
+template <FieldType fieldType, bool isArithmetic>
+struct ArithmeTicFieldtypeTraits : public BaseFieldtypeTraits<fieldType> {
```
Comment (chaoyli) [inline be/src/olap/types.h:541]: Numeric may be a better name.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/types.h (hunk 3 above)
Comment (HappenLee) [reply be/src/olap/types.h:541]: OK，i will consider to change the name of it

File: be/src/olap/types.h (hunk 2 above)
Comment (HappenLee) [reply be/src/olap/types.h:563]: Now， ArithmeticField just to keep the small with old code.
old code only Partial Template Specialization the OLAP_FIELD_TYPE_INT but not OLAP_FIELD_TYPE_UNSIGNED_INT. There is no need to Partial Template Specialization for unsign type, so there check the CppType whether is_signed

File: be/src/olap/types.h (hunk 1 above)
Comment (HappenLee) [reply be/src/olap/types.h:514]: Hi, chaoyli.

```if constexpr``` only support in c++17 and ```std::is_floating_point_v``` support in c++14.

Now, Doris only support C++11, So I use label assignment to distinguish the float and int.

**Review by chaoyli — APPROVED** [review]

**Pull request merged**